File: 46. Permutations/46. Permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
  const n = nums.length;
  const ans = [];
  const curr = [];
  const used = new Array(n).fill(false);

  const backtrack = () => {
    if (curr.length === n) {
      ans.push([...curr]);
      return;
    }

    for (let i = 0; i < n; i++) {
      if (!used[i]) {
        curr.push(nums[i]);
        used[i] = true;
        backtrack();
        curr.pop();
        used[i] = false;
      }
    }
  };

  backtrack();
  return ans;
};

File: 46. Permutations/46. Permutations.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ans, curr = [], []
        used = [False] * n

        def backtrack():
            if len(curr) == n:
                ans.append(curr[:])
                return

            for i in range(n):
                if not used[i]:
                    used[i] = True
                    curr.append(nums[i])
                    backtrack()
                    curr.pop()
                    used[i] = False

        backtrack()
        return ans
